examples/reshare/src/dkg/manager.rs

```diff
@@ -0,0 +1,425 @@
+        primitives::{group::Share, poly::Public, variant::Variant},
+    },
+    Hasher, PrivateKey, PublicKey, Verifier,
+};
+use commonware_p2p::{Receiver, Recipients, Sender};
+use futures::FutureExt;
+use rand_core::CryptoRngCore;
+use std::collections::HashMap;
+use tracing::{debug, error, info, warn};
+
+/// The signature namespace for DKG acknowledgment messages.
+const ACK_NAMESPACE: &[u8] = b"COMMONWARE_DKG_ACK";
+
+/// The concurrency level for DKG/reshare operations.
+const CONCURRENCY: usize = 1;
+
+/// A manager for a DKG/reshare round.
+///
+/// Exposes functionality:
+/// - Distribute a [Dealer]'s shares
+/// - Process incoming shares and acknowledgements from other [Dealer]s and [Player]s.
+/// - Create a [DealOutcome] from the current state, for inclusion in a [Block].
+/// - Process [Block]s that may contain a [DealOutcome].
+/// - Finalize the DKG/reshare round, returning the resulting [Output].
+pub struct DkgManager<'ctx, V, P, S, R>
+where
+    V: Variant,
+    P: PrivateKey,
+    S: Sender<PublicKey = P::PublicKey>,
+    R: Receiver<PublicKey = P::PublicKey>,
+{
+    /// The local signer.
+    signer: &'ctx mut P,
+
+    /// The index of the local signer in the contributors list.
+    signer_index: usize,
+
+    /// The previous group polynomial and share.
+    previous: Output<V>,
+
+    /// The contributors to the round.
+    contributors: &'ctx [P::PublicKey],
+
+    /// The outbound communication channel for peers.
+    sender: &'ctx mut S,
+
+    /// The inbound communication channel for peers.
+    receiver: &'ctx mut R,
+
+    /// The local dealer for this round.
+    dealer: Dealer<P::PublicKey, V>,
+
+    /// The dealer's commitment.
+    commitment: Public<V>,
+
+    /// The dealer's shares for all players.
+    shares: Vec<Share>,
+
+    /// The local player for this round.
+    player: Player<P::PublicKey, V>,
+
+    /// The local arbiter for this round.
+    arbiter: Arbiter<P::PublicKey, V>,
+
+    /// Signed acknowledgements from contributors.
+    acks: HashMap<u32, P::Signature>,
+
+    /// The deal outcome constructed from this manager's state, if any.
+    deal_outcome: Option<DealOutcome<P, V>>,
+}
+
+impl<'ctx, V, P, S, R> DkgManager<'ctx, V, P, S, R>
+where
+    V: Variant,
+    P: PrivateKey,
+    S: Sender<PublicKey = P::PublicKey>,
+    R: Receiver<PublicKey = P::PublicKey>,
+{
+    /// Create a new DKG/reshare manager.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the signer is not in the list of contributors.
+    pub fn new<E: CryptoRngCore>(
+        context: &mut E,
+        public: Public<V>,
+        share: Share,
+        signer: &'ctx mut P,
+        contributors: &'ctx [P::PublicKey],
+        sender: &'ctx mut S,
+        receiver: &'ctx mut R,
+    ) -> Self {
+        let signer_index = contributors
+            .iter()
+            .position(|pk| pk == &signer.public_key())
+            .expect("signer must be in contributors");
+
+        let (dealer, commitment, shares) =
```

Comment:
We haven't discussed this yet but we need to persist this information prior to distributing it (we derive it with OsRng and shouldn't rederive it upon restart).